benchmark/mnist_benchmark.py

```diff
@@ -0,0 +1,668 @@
+        "final_test_accuracy": final_metrics["accuracy"],
+    }
+
+
+def pca(x: jnp.ndarray, n_components: int = 16) -> jnp.ndarray:
+    """
+    Perform Principal Component Analysis (PCA) on the dataset x.
+
+    This function computes the principal components of the input data matrix
+    and returns the projected data.
+
+    :param x: The input data matrix of shape (n_samples, n_features).
+    :param n_components: The number of principal components to return.
+    :return: The projected data of shape (n_samples, n_components).
+    """
+    # Center the data
+    x_centred = x - jnp.mean(x, axis=0)
+
+    # Compute the covariance matrix
+    cov_matrix = jnp.cov(x_centred.T)
+
+    # Compute eigenvalues and eigenvectors
+    eigenvalues, eigenvectors = jnp.linalg.eigh(cov_matrix)
+
+    # Sort eigenvectors by descending eigenvalues
+    _idx = jnp.argsort(eigenvalues)[::-1]
+    eigenvectors = eigenvectors[:, _idx]
+
+    # Select top n_components eigenvectors
+    components = eigenvectors[:, :n_components]
+
+    # Project the data onto the new subspace
+    x_pca = jnp.dot(x_centred, components)
+
+    return x_pca
+
+
+def prepare_datasets() -> tuple:
+    """
+    Prepare and return training and test datasets.
+
+    :return: A tuple containing training and test datasets in JAX arrays:
+             (train_data_jax, train_targets_jax, test_data_jax, test_targets_jax).
+    """
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Lambda(lambda x: x.view(-1))]
+    )
+    train_dataset = torchvision.datasets.MNIST(
+        root="./data", train=True, download=True, transform=transform
+    )
+    test_dataset = torchvision.datasets.MNIST(
+        root="./data", train=False, download=True, transform=transform
+    )
+
+    train_data_jax, train_targets_jax = convert_to_jax_arrays(train_dataset)
+    test_data_jax, test_targets_jax = convert_to_jax_arrays(test_dataset)
+
+    return (train_data_jax, train_targets_jax, test_data_jax, test_targets_jax)
+
+
+def initialise_solvers(train_data_pca: jnp.ndarray, key: jax.random.PRNGKey) -> list:
+    """
+    Initialise and return a list of solvers for various coreset algorithms.
+
+    Set up solvers for Kernel Herding, Stein Thinning, Random Sampling, and Randomised
+    Cholesky methods. Each solver has different parameter requirements. Some solvers
+    can utilise MapReduce, while others cannot,and some require specific kernels.
+    This setup allows them to be called by passing only the coreset size,
+    enabling easy integration in a loop for benchmarking.
+
+    :param train_data_pca: The PCA-transformed training data used for
+                           length scale estimation for ``SquareExponentialKernel``.
+    :param key: The random key for initialising random solvers.
+
+    :return: A list of solvers functions for different coreset algorithms.
+    """
+    # Set up kernel using median heuristic
+    num_samples_length_scale = min(300, 1000)
```

Comment:
Did you mean, as exists elsewhere in the codebase, ` num_samples_length_scale = min(num_data_points, 1_000) `? `Where num_data_points = len(train_data_pca)`